### PR TITLE
feat: introducing CancellationTokenRegistry

### PR DESCRIPTION
**Pull request by axel7083**
### What does this PR do?

This PR is adding a _very similar_ CancellationTokenRegistry. It is using the class CancellationTokenSource from the api to do so.

### What issues does this PR fix or reference?

Fixes https://github.com/containers/podman-desktop-extension-ai-lab/issues/724

### How to test this PR?

- [x] unit tests has been provided

### CODE REVIEW

**Review by lstocchi — COMMENTED** [review]
Missing a function to create a new cancel token?
i see you just created one to ask the cancellation of an existing token

File: packages/backend/src/registries/CancellationTokenRegistry.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+export class CancellationTokenRegistry implements Disposable {
+  #callbackId: number;
+  #callbacksCancellableToken: Map<number, ICancellationTokenSource>;
+
+  constructor() {
+    this.#callbackId = 0;
+    this.#callbacksCancellableToken = new Map<number, ICancellationTokenSource>();
+  }
+
+  createCancellationTokenSource(): number {
```
Comment (axel7083) [inline packages/backend/src/registries/CancellationTokenRegistry.ts:30]: @lstocchi

Comment (axel7083) [reply packages/backend/src/registries/CancellationTokenRegistry.ts:30]: This are the same methods use in PodmanDesktop, I wanted to stay close to their implementation

Comment (lstocchi) [reply packages/backend/src/registries/CancellationTokenRegistry.ts:30]: I express myself badly. I was referring to the studioImpl. You only exposed the cancelToken but not the create. I would expect a `requestNewToken  `https://github.com/containers/podman-desktop-extension-ai-lab/pull/725/files#diff-3f3c50af66ddba22163580bcd3807fbb24558686b62bebd09e120bc6be130a76R394-R398

Comment (axel7083) [reply packages/backend/src/registries/CancellationTokenRegistry.ts:30]: IMO the frontend should not be responsible of creating those token, but should be attached to a task. 

Let's take the download model example, the `ModelManager` will create a cancellation token, set to the download task a `cancellationId` that the frontend will be able to see. But there is no reason why the front should create those tokens

Comment (lstocchi) [reply packages/backend/src/registries/CancellationTokenRegistry.ts:30]: I'm not so convinced but we can always add it later 👍

**Review by lstocchi — APPROVED** [review]
LGTM

**Review by benoitf — COMMENTED** [review]
AFAIK `CancellationTokenSource` should be available from the extension API

(CancellationToken, CancellationTokenSource)

**Review by benoitf — COMMENTED** [review]
basically CancellationTokenSource should be exposed as a class, not an interface from the Podman Desktop API

**Comment by axel7083** [comment]
> AFAIK `CancellationTokenSource` should be available from the extension API
> 
> (CancellationToken, CancellationTokenSource)

It is available, but as I mention, as an interface not a class, is this what you are pointing out ? It should be available as a class ?

**Comment by axel7083** [comment]
> basically CancellationTokenSource should be exposed as a class, not an interface from the Podman Desktop API

Okey, I will open a ticket on the Podman Desktop side, but for our timeframe and the version we support, I guess we cannot rely on the api to handle this for us

**Comment by benoitf** [comment]
> Okey, I will open a ticket on the Podman Desktop side, but for our timeframe and the version we support, I guess we cannot rely on the api to handle this for us

I don't know for timeframe, there is a release planned April 2nd (1.9.0 of Podman Desktop)

**Comment by axel7083** [comment]
> > Okey, I will open a ticket on the Podman Desktop side, but for our timeframe and the version we support, I guess we cannot rely on the api to handle this for us
> 
> I don't know for timeframe, there is a release planned April 2nd (1.9.0 of Podman Desktop)

@slemeur do we want to make `AI-Lab` only compatible with 1.9 ?

**Review by benoitf — APPROVED** [review]

**Pull request merged**